src/irmin-pack/IO.ml

```diff
@@ -117,28 +117,50 @@ module Unix : S = struct
 
   type t = {
     file : string;
-    mutable raw : Raw.t;
+    mutable raw : Raw.t option;
     mutable offset : int64;
     mutable flushed : int64;
     readonly : bool;
     version : string;
     buf : Buffer.t;
   }
 
+  let may f = function None -> () | Some bf -> f bf
+
+  let assert_and_get = function None -> assert false | Some e -> e
+
+  (* only RO instances could not have the file opened yet. *)
+  let try_load_file t =
+    match t.raw with
+    | Some _ -> ()
+    | None ->
+        Log.debug (fun l -> l "checking on-disk %s" t.file);
+        if Sys.file_exists t.file then (
+          let x =
+            Unix.openfile t.file Unix.[ O_EXCL; O_RDONLY; O_CLOEXEC ] 0o644
+          in
+          let raw = Raw.v x in
+          let version = Raw.unsafe_get_version raw in
+          assert (version = t.version);
```

Comment:
I assume this was taken from elsewhere, but the use of `assert` is suspicious. Isn't this visible to the user if the version is changed concurrently?

Reply:
I think that the point of having a `version`, is to ensure that only files compatible with the current version of pack are read. It should not be possible to change it "on-the-fly".

Reply:
If a user upgrades `Irmin_pack` and tries to read an old file, will they see an `assert` failure here? My thinking was that we might want a `Corrupted` or `Unsupported_version` exception here instead, rather than putting such checks in `asserts` where (a) they can be compiled away and (b) they look like constraints that have already been ensured by the code (which I don't think is the case here).

Reply:
Ah, ok, I get your point now (the concurrent part confused me, sorry). A more informative failure is definitely better, I'll add it!